rest/route/keys_routes.go

```diff
@@ -52,3 +67,70 @@ func (h *keysGetHandler) Execute(ctx context.Context, sc data.Connector) (Respon
 		Result: userKeys,
 	}, nil
 }
+
+type keysPostHandler struct {
+	keyName  string
+	keyValue string
+}
+
+func (h *keysPostHandler) Handler() RequestHandler {
+	return &keysPostHandler{}
+}
+
+func (h *keysPostHandler) ParseAndValidate(ctx context.Context, r *http.Request) error {
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	key := model.APIPubKey{}
+	if err := util.ReadJSONInto(body, &key); err != nil {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    fmt.Sprintf("failed to unmarshal public key: %s", err),
+		}
+	}
+
+	if err := h.validatePublicKey(key); err != nil {
+		return &rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    fmt.Sprintf("invalid public key: %s", err),
+		}
+	}
+
+	return nil
+}
+
+// XXX: If you are changing the validation in this function, you must also
+// update keyBaseValid in directives.spawn.js
+func (h *keysPostHandler) validatePublicKey(key model.APIPubKey) error {
+	h.keyName = string(key.Name)
+	h.keyValue = string(key.Key)
+	if strings.TrimSpace(h.keyName) == "" {
+		return errors.New("empty key name")
+	}
+
+	if !strings.HasPrefix(h.keyValue, "ssh-rsa") || !strings.HasPrefix(h.keyValue, "ssh-dss") {
+		return errors.New("invalid public key")
+	}
+
+	splitKey := strings.Split(h.keyValue, " ")
+	if len(splitKey) < 2 {
+		return errors.New("invalid public key")
+	}
+
+	matched, err := regexp.MatchString(keyRegex, splitKey[1])
+	if err != nil || !matched {
+		return errors.New("invalid public key")
+	}
+
+	return nil
+}
+
+func (h *keysPostHandler) Execute(ctx context.Context, sc data.Connector) (ResponseData, error) {
+	u := MustHaveUser(ctx)
+
+	if err := sc.AddPublicKey(u, h.keyName, h.keyValue); err != nil {
+		return ResponseData{}, errors.Wrap(err, "failed to add key")
+	}
+
+	return ResponseData{}, nil
```

Comment:
Not a huge deal, but I would return some sort of body in the response here (maybe the key name and a confirmation message?)